app/scripts/modules/core/src/presentation/forms/fields/FormikFormField.tsx

```diff
@@ -70,13 +82,56 @@ function FormikFormFieldImpl<T = any>(props: IFormikFormFieldImplProps<T>) {
     messageNode,
   };
 
+  const [inputMode, setInputMode] = React.useState(SpelAwareInputMode.DEFAULT);
+
+  const modeToInput: { [k in SpelAwareInputMode]: React.ComponentType<IFormInputProps> } = {
+    [SpelAwareInputMode.DEFAULT]: input,
+    [SpelAwareInputMode.FREEFORM]: SimpleSpelInput,
+  };
+
+  const freeformInputEnabled = firstDefined(
+    get(props, 'spelConfig.freeformInputEnabled', null),
+    get(useContext(FormikSpelContext), 'freeformInputEnabled', null),
+    false,
+  );
+
+  React.useEffect(() => {
+    if (!freeformInputEnabled) {
+      return;
+    }
+    const fieldValue = get(props.formik.values, name, '');
+    const isFieldValueSpel = SpelService.includesSpel(fieldValue);
+    if (isFieldValueSpel) {
+      setInputMode(SpelAwareInputMode.FREEFORM);
+    }
+  }, []);
+
+  const onSpelToggleClick = () => {
+    formik.setFieldValue(name, null);
+    setInputMode(inputMode === SpelAwareInputMode.DEFAULT ? SpelAwareInputMode.FREEFORM : SpelAwareInputMode.DEFAULT);
+  };
+
   const renderField = ({ field }: FieldProps<any>) => {
-    const inputRenderPropOrNode = firstDefined(input, noop);
+    const inputRenderPropOrNode = firstDefined(modeToInput[inputMode], noop);
```

Comment:
I feel like this might be simpler if we co-located the default/freeform lookup with its usage and nuked the modeToInput mapping
```suggestion
    const inputOrSpel = inputMode === SpelAwareInputMode.DEFAULT ? input : SimpleSpelInput;
    const inputRenderPropOrNode = firstDefined(inputOrSpel, noop);
```
dis/agree?

If the intention is that the map can have additional modes added to it, can we use a function instead?

Reply:
Agree, thank you! Also nuked the mode to message mapping in `SpelToggle`.